gips/0046-graphql-validations.md

```diff
@@ -0,0 +1,108 @@
+- `VariablesInAllowedPosition` (https://spec.graphql.org/draft/#sec-All-Variable-Usages-Are-Allowed)
+- `KnownDirectives` (https://spec.graphql.org/draft/#sec-Directives-Are-Defined)
+- `UniqueDirectivesPerLocation` (https://spec.graphql.org/draft/#sec-Directives-Are-Unique-per-Location)
+
+The following spec rules are not part of this implementation:
+
+- `ExecutableDefinitions` - not needed, because of `graphql_parser` nature.
+- `UniqueInputFieldNames` (blocked by https://github.com/graphql-rust/graphql-parser/issues/59, low priority for `graph-node`)
+
+In case of a validation error as part of the `validate` phase, `graph-node` will return a validation error, and will skip the GraphQL execution:
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-1.png)
+
+![GraphQL Validation Error](../assets/gip-0046/graphql-validations-2.png)
+
+# Backwards Compatibility
+
+The GraphQL execution layer will run in the same way, but due to the nature of the extra added validation checks, some users might experience a change in the way `graph-node` responds to end-users GraphQL operations.
+
+To address the potential breaking change for users, we applied the following measures:
+
+1. Document the change, how it might effect consumers and how to migrate/fix invalid operations (https://github.com/graphprotocol/docs/pull/284)
+2. Provide a CLI tool that can locate potential operations that might be invalid at runtime (documented in https://github.com/graphprotocol/docs/pull/284)
+3. Find high-traffic and high-profile Subgraphs and contact the owners, in order to give them enough time to fix invalid operations.
+4. Run GraphQL validations as part of the Studio GraphiQL editor, to make sure users are aware of this change ahead of time (https://github.com/edgeandnode/graphiql-playground/pull/17).
+5. Announce GraphQL validations is coming, and provide enough time for consumers to adjust.
+
+Due to the nature of this change, only end-users and not Subgraph developers are effected. We are not able to find all effected users, but we can give users enough time to adjust and fix their invalid GraphQL operations.
+
+# Dependencies
+
+A single dependency on an external package [`graphql-tools-rs`](https://github.com/dotansimha/graphql-tools-rs) (developed and maintained by The Guild) is required. This package performs and implements the actual GraphQL validation phase. This package also runs all checks and validations, as defined in the GraphQL reference implementation and by the GraphQL specification.
+
+# Risks and Security Considerations
+
+A performance risk is in place, due to the nature of adding a new phase to the GraphQL execution pipeline. It as mitigated by code review by `graph-node` project leads. All changes were tested on standalone environments prior to rollout. In In addition, the production environment is running the GraphQL validations phase in silent mode
```

Comment:
I would emphasize running it in silent mode rather than the code review, as that is what really gives us confidence this won't have an impact (as the validations are effectively "active" already, just not acted upon)